### PR TITLE
fix Chunk2DocRankDriver 763

### LINKED ISSUE #763: MinRanker is not using the proper driver
**Describe the bug**
`MinRanker` should use `Chunk2DocRankDriver` instead of `BaseRankDriver`. The `BaseRankDriver` does not merge the chunk level information to the doc level

**Describe how you solve it**

- - -

**Environment**

**Screenshots**

┆Issue is synchronized with this [Jira Task](https://jinaai.atlassian.net/browse/JINACORE-145) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by nan-wang**
resolves #763

### CODE REVIEW

File: jina/drivers/__init__.py [hunk 1]
```diff
@@ -258,7 +258,7 @@ def pre_traverse(_docs, traverse_on, context_doc=None):
             _traverse(docs, 'chunks')
         if 'matches' in self.traverse_fields:
             for d in docs:
-                _traverse(d.matches, 'matches')
+                _traverse(d, 'matches')
```
Comment (hanxiao) [inline jina/drivers/__init__.py:261]: no this is wrong, please revert it back.

Comment (hanxiao) [reply jina/drivers/__init__.py:261]: _traverse do not work on single doc

Comment (nan-wang) [reply jina/drivers/__init__.py:261]: yes, u r right.

**Comment by hanxiao** [comment]
is this ready for review? @nan-wang

File: jina/resources/executors.requests.Chunk2DocRanker.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+on:
+  ControlRequest:
+    - !ControlReqDriver {}
+  SearchRequest:
+    - !Chunk2DocRankDriver
+      with:
+        depth_range: [0, 2]  # prune root and 1st level
+        start: 0
+        end: 50
```
Comment (hanxiao) [inline jina/resources/executors.requests.Chunk2DocRanker.yml:9]: what are these start and end is about?

Comment (hanxiao) [inline jina/resources/executors.requests.Chunk2DocRanker.yml:9]:
```suggestion

```

Comment (hanxiao) [reply jina/resources/executors.requests.Chunk2DocRanker.yml:9]: and is there any reason why you want to promote a specialized executor to `resources` level?

**Comment by JoanFM** [comment]
Enables https://github.com/jina-ai/examples/pull/132

**Comment by nan-wang** [comment]
> is this ready for review? @nan-wang 

not yet. This is for fixing failure of retrieving docs in examples.

File: tests/unit/drivers/test_rank_driver.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+class Chunk2DocRankDriverTestCase(JinaTestCase):
+    def test_chunk2doc_driver_mock_indexer(self):
+        doc = create_document_to_rank()
+        print(doc)
+        driver = Chunk2DocRankDriver()
+        executor = MockRanker()
+        args = set_pea_parser().parse_args(['--runtime', 'process'])
+        pea = MockPea(args)
+
+        driver.attach(executor=executor, pea=pea)
```
Comment (JoanFM) [inline tests/unit/drivers/test_rank_driver.py:91]: You can safely set pea to None I think and avoid having a MockPea.

Comment (nan-wang) [reply tests/unit/drivers/test_rank_driver.py:91]: Nope, the driver needs to check the `self.pea.message.envelope.status.code` in order to decide on which level to skip the execution
https://github.com/jina-ai/jina/blob/master/jina/drivers/__init__.py#L293

File: tests/unit/drivers/test_rank_driver.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+class MockPea(BasePea):
+    def __init__(self, args):
+        super().__init__(args)
+        self._message = jina_pb2.Message()
+        self._message.envelope.status.code = jina_pb2.Status.SUCCESS
+
+
+class Chunk2DocRankDriverTestCase(JinaTestCase):
+    def test_chunk2doc_driver_mock_indexer(self):
```
[the remark this answers is not in the document]
Comment (nan-wang) [reply tests/unit/drivers/test_rank_driver.py:83]: fair enough. I'll change it.

File: tests/unit/drivers/test_rank_driver.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+from tests import JinaTestCase
```
[the remark this answers is not in the document]
Comment (nan-wang) [reply tests/unit/drivers/test_rank_driver.py:1]: hmm, I've just found it. `test_chunk2doc_rank_drivers.py`.

File: jina/drivers/rank.py [hunk 6]
```diff
@@ -28,11 +29,12 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.recursion_order = 'post'
 
-    def _apply_all(self, docs: 'jina_pb2.Document', context_doc: 'jina_pb2.Document', *args, **kwargs):
+    def _apply_all(self, docs: List['jina_pb2.Document'], context_doc: 'jina_pb2.Document', travers_on: str, *args, **kwargs):
```
Comment (hanxiao) [inline jina/drivers/rank.py:32]: if you don't need `traverse_on`, no need to add it. That's how *args and **kwargs for

File: jina/drivers/rank.py [hunk 7]
```diff
@@ -28,11 +29,12 @@ def __init__(self, *args, **kwargs):
...
         """
 
         :param docs: the chunks of the ``context_doc``, they are at depth_level ``k``
         :param context_doc: the owner of ``docs``, it is at depth_level ``k-1``
+        :param traverse_on: "matches" or "chunks"
```
Comment (hanxiao) [inline jina/drivers/rank.py:37]: same here, if you don't need in the function, no need to add it

File: jina/drivers/__init__.py [hunk 8]
```diff
@@ -254,6 +255,7 @@ def pre_traverse(_docs, traverse_on, context_doc=None):
         else:
             raise ValueError(f'{self.recursion_order}')
 
+        # context_doc is missing, and this leads to `_apply_all` is skipped
```
Comment (hanxiao) [inline jina/drivers/__init__.py:258]: what does this mean? can you remove this comment if nothing wrong with the logic?

also, always refer to #702 illustration about the doc structure

File: jina/drivers/rank.py (hunk 7 above)
Comment (nan-wang) [reply jina/drivers/rank.py:37]: Yes, u r right. The codes need clean up.

File: jina/drivers/__init__.py (hunk 8 above)
Comment (nan-wang) [reply jina/drivers/__init__.py:258]: Some debug messages left over. As above, it has been cleaned up.

File: tests/test_helper.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+from tests import JinaTestCase
+from jina.helper import cached_property
+
+
+class MyTestCase(JinaTestCase):
```
Comment (JoanFM) [inline tests/test_helper.py:5]: can you put a more descriptive name to the `TestCase` class? And shouldn't it be under `tests/unit`?

File: tests/test_helper.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
+from tests import JinaTestCase
```
Comment (JoanFM) [inline tests/test_helper.py:1]: I think these tests are a great idea, we should also add a `test` for `batching` decorator

File: jina/helper.py [hunk 11]
```diff
@@ -647,6 +647,6 @@ def __init__(self, func):
     def __get__(self, obj, cls):
         if obj is None:
             return self
-
-        value = obj.__dict__[f'CACHED_{self.func.__name__}'] = self.func(obj)
```
Comment (JoanFM) [inline jina/helper.py:651]: In `jina/executors/__init__.py` there is some logic related to this that needs to be reviewed maybe?

```python
    def __getstate__(self):
        d = dict(self.__dict__)
        del d['logger']
        for k in self._post_init_vars:
            del d[k]
        cached = [k for k in d.keys() if k.startswith('CACHED_')]
        for k in cached:
            del d[k]
        return d
```

Comment (JoanFM) [reply jina/helper.py:651]: or was it the source of the problem? that we should say `if not  k.startswith('CACHED_')`?

Comment (nan-wang) [reply jina/helper.py:651]: Good catch! This part of the codes is for dumping `executors` into pickles. We don't want to save the cached variables, e.g. the models. However, using the `CACHED_` prefix leads to the failure of retrieving the variable from the `self.__dict__`.

Comment (hanxiao) [reply jina/helper.py:651]: your changes will fail. because you save cached which are not pickable. please revert it back

**Comment by hanxiao** [comment]
@nan-wang I'm confused, what is this PR really about? As far as I can tell, it adds some unit tests. problem reported in #763 is either not-an-issue or this hasn't resolved in this PR yet. And now it tries to resolve `cached_property` without supported unit tests.

## Suggestions

- change the name of this PR, label  #763 with correct status; 
- resolve cached_property separately in another PR with supported unit test.

File: jina/drivers/__init__.py [hunk 12]
```diff
@@ -260,6 +260,10 @@ def pre_traverse(_docs, traverse_on, context_doc=None):
         if 'matches' in self.traverse_fields:
             for d in docs:
                 _traverse(d.matches, 'matches')
+        if 'chunk_matches' in self.traverse_fields:
```
Comment (JoanFM) [inline jina/drivers/__init__.py:263]: I think this effect is already achieved by adding chunks, matches. Given that is if, if, first will iterate on chunks and later on matches.

Comment (hanxiao) [reply jina/drivers/__init__.py:263]: @JoanFM is right. no need of this logic . simply set traverse_on to [chunks, matches] does this job.

File: jina/helper.py (hunk 11 above)
Comment (JoanFM) [reply jina/helper.py:677]: This has been handled by another PR

File: jina/drivers/__init__.py (hunk 12 above)
Comment (hanxiao) [reply jina/drivers/__init__.py:263]: https://github.com/jina-ai/jina/pull/757#issuecomment-667923756

plz see my previous comment.

btw, this PR is really sidetracked and takes too long time. i strongly suggested to close this PR with your added unittest only, if there are fixes on Transformers then started another PR.

or at least change the title of this PR or the issue description #762

**Comment by nan-wang** [comment]
rephrased in #799

**Pull request closed without merging**